machine-learning/src/main.py

```diff
@@ -14,59 +17,87 @@
 clip_text_model = os.getenv('MACHINE_LEARNING_CLIP_TEXT_MODEL', 'clip-ViT-B-32')
 
 _model_cache = {}
+_executor = ThreadPoolExecutor(max_workers=10)
+
+
 def _get_model(model, task=None):
-  global _model_cache
-  key = '|'.join([model, str(task)])
-  if key not in _model_cache:
-    if task:
-      _model_cache[key] = pipeline(model=model, task=task)
+    global _model_cache
+    key = '|'.join([model, str(task)])
+    if key not in _model_cache:
+        if task:
+            _model_cache[key] = pipeline(model=model, task=task)
+        else:
+            _model_cache[key] = SentenceTransformer(model)
+        # set a timeout of 1 hour for the model object
+        _model_cache[key + '|timestamp'] = datetime.now() + timedelta(hours=1)
     else:
-      _model_cache[key] = SentenceTransformer(model)
-  return _model_cache[key]
+        # check if the model object has expired
+        timestamp = _model_cache.get(key + '|timestamp')
+        if timestamp and datetime.now() > timestamp:
+            del _model_cache[key]
+            del _model_cache[key + '|timestamp']
```

Comment:
When I was testing earlier, I deleted the model from the cache, but the memory usage did not go down. Did you observe something different?

Reply:
No I didn't see really chance but I also did not let it run for over an hour. I also think that the whole code needs to be profiled to figure out which portion is the largest memory hog (I suspect the model cache that wasn't being shared previously between workers). I'm going to try to do that tomorrow.

I was originally trying to use multiprocessing but couldn't quite get it to work with the clip models.

Reply:
Why don't you change it to one minute and test? 😁

I deleted it immediately and it never went down... So idk lol.

Reply:
You might need to call Python's garbage collector to release the memory, otherwise it might take a while until it is released. 

```
import gc
gc.collect()
```

Reply:
I instantiated the model, then deleted the object, then called gc collect and memory usage remain unchanged.

Reply:
I gave up at that point. I assume there is something going on in the underlying libraries with how they manage loading models. They might have their own cache, etc. Would probably require more research to figure it out, but I moved on to other things 😛.

Reply:
I gave up at that point. I assume there is something going on in the underlying libraries with how they manage loading models. They might have their own cache, etc. Would probably require more research to figure it out, but I moved on to other things 😛.